### PR TITLE
Microdata Improvements

### PR DESCRIPTION
**Pull request by jrbenito**
### Configuration

 - **Operating system with version** :  Ubuntu (tested on Chrome Windows)
 - **Node version** :  4.3.3
 - **Hexo version** :  3.2.2
 - **Hexo-cli version** :  
 
### Changes proposed

 - Improve microdata (BlogPosting) on post page
 - Improve microdata (BlogPosting) on main index page
 - Add microdata (Person) on sidebar (author)

This improvements silence some warning and errors from [Google Structured Data Testing Tool](https://search.google.com/structured-data/testing-tool).

### Test Method

Use Google´s testing tool before and after this patch to see improvements. 

### Expected Results

With improvements on microdata Google Search Engine (and hopefully other services) might create reach results (reach cards in case of Google, for instance).

### CODE REVIEW

File: layout/_partial/index.ejs [hunk 1]
```diff
@@ -32,8 +32,14 @@
...
                 <a href="<%- url_for(post.path) %>">
-                    <div class="postShorten-thumbnailimg">
-                        <img alt="" itemprop="image" src="<%= thumbnailImageUrl %>"/>
+                    <div class="postShorten-thumbnailimg" itemprop="image" itemscope itemtype="https://schema.org/ImageObject">
+                        <img alt="" src="<%= thumbnailImageUrl %>"/>
+                        <meta itemprop="url" content="<%= thumbnailImageUrl %>">
+                        <% // width and height are required ImageObject, values hardcode should have way to populate from
+                           // css variables 
+                        %>
+                        <meta itemprop="width" content="140">
```
Comment (jrbenito) [inline layout/_partial/index.ejs]: 140 px was derived from CSS files. This shall be the same for display and metadata, I did not figure out a way to grab information from css and throw it here dynamically during theme build. This would be ideal to avoid mismatches in future. However, it might not harm.

File: layout/_partial/post/meta.ejs [hunk 2]
```diff
@@ -6,8 +6,14 @@
...
     	<% } %>
     </time>
+    <% if (post.updated) { %>
+    <time itemprop="dateModified" datetime="<%= post.updated.format() %>" />
+    <% } %>
     <% if ((post.categories) && (post.categories.length > 0)) { %>
         <span><%= __('post.categorized_in')  %> </span>
         <%- partial('category', {categories: post.categories})%>
     <% } %>
+    <div itemprop="author" itemscope itemtype="http://schema.org/Person">
```
Comment (LouisBarranqueiro) [inline layout/_partial/post/meta.ejs]: author name is already set in the `head` tag of the page

File: layout/_partial/index.ejs [hunk 3]
```diff
@@ -88,8 +94,14 @@
         </div>
         <% if (thumbnailImageUrl !== null && thumbnailImagePosition !== "top") { %>
             <a href="<%- url_for(post.path) %>">
-                <div class="postShorten-thumbnailimg">
-                    <img alt="" itemprop="image" src="<%= thumbnailImageUrl %>"/>
+                <div class="postShorten-thumbnailimg" itemprop="image" itemscope itemtype="https://schema.org/ImageObject">
```
Comment (LouisBarranqueiro) [inline layout/_partial/index.ejs]: same thing here

File: layout/_partial/index.ejs (hunk 1 above)
Comment (LouisBarranqueiro) [inline layout/_partial/index.ejs]: It will be hard to maintain with this hardcoded values. Open graph data already do this job so we don't really need it.

File: layout/_partial/sidebar.ejs [hunk 4]
```diff
@@ -27,9 +27,9 @@
             <% for (var n in theme.sidebar[i]) { %>
                 <li class="sidebar-button">
                     <% if (url_for(theme.sidebar[i][n].url).indexOf(config.url) < 0 && url_for(theme.sidebar[i][n].url).indexOf(':') >= 0) { %>
-                        <a  class="sidebar-button-link <% if (theme.sidebar[i][n].class) { %><%= theme.sidebar[i][n].class %><% } %>" href="<%- url_for(theme.sidebar[i][n].url) %>" target="_blank" rel="noopener">
+                        <a itemprop="sameAs" class="sidebar-button-link <% if (theme.sidebar[i][n].class) { %><%= theme.sidebar[i][n].class %><% } %>" href="<%- url_for(theme.sidebar[i][n].url) %>" target="_blank">
```
Comment (LouisBarranqueiro) [inline layout/_partial/sidebar.ejs]: Why did you remove `rel="noopener"`?

File: layout/_partial/index.ejs (hunk 1 above)
Comment (jrbenito) [inline layout/_partial/index.ejs]: In fact it is needed for google to correct try create rich cards and other things about a site. Their tool complain about it. If you want,  I remove it from PR.

File: layout/_partial/post/meta.ejs (hunk 2 above)
Comment (jrbenito) [inline layout/_partial/post/meta.ejs]: the head tag does not apply to "article" schema.org. Again, Google´s structured testing tool will complain that every single entity of the page be marked even it being redundant. As per webmaster discussions there, in the index page for instance, every single excerpt  is an "BlogPosting" entity and a BlogPosting must explicitly define its Author (@person), Publisher (@Organization for google´s purpose, so I left it out for now because it yelds other errors if set to @Person). and image (I set it optionally if the posting image is defined by user - as theme already do in fact).

File: layout/_partial/sidebar.ejs (hunk 4 above)
Comment (jrbenito) [inline layout/_partial/sidebar.ejs]: Sorry, merge error on my side, will correct it.

**Comment by jrbenito** [comment]
Just to clarify my intent here, this PR tries to make BlogPosting schema more compliant with google standards. The hope is to google correct use this information for creating rich cards and carousels for the posts when relevant.

If you think this is not important add on, feel free to discard.

File: layout/_partial/sidebar.ejs (hunk 4 above)
Comment (jrbenito) [inline layout/_partial/sidebar.ejs]: This happened because I did edit files on my blog for test purposes and, there, I am using v1.9.1 so, differences between there and dev branch are expected. However, I did not pay attention that, if I cherry pick files from there to my theme fork, git would "subtract" differences (which make sense a lot).

**Comment by jrbenito** [comment]
Rebase over DEV, added rel=... back

**Comment by LouisBarranqueiro** [comment]
> Just to clarify my intent here, this PR tries to make BlogPosting schema more compliant with google standards. The hope is to google correct use this information for creating rich cards and carousels for the posts when relevant.

I understand but what are the real benefits of this changes? If we add some hardcoded values and increase complexity of the code just to silent errors without any real benefits behind that, it not worth it. For example, with Open Graph, if I add a tag or another, I can see the result with this [tool](https://developers.facebook.com/tools/debug/). So if you can show me that the card is richer with the changes of this PR, I'm ok to merge it.

**Comment by jrbenito** [comment]
Hi Louis,

Based on Google´s documentation, if there is errors while checking by their tool, the search engine will most likely ignore schema tagging and won´t form rich cards (or rich snippets) of the site. However, I think google is way more obscure than Facebook OG. As told, the goal is to make Google consider embedded microdata but I understand your concern.

The rich snippets, as you may know, is like [this](https://www.google.com.br/search?q=google+rich+cards&espv=2&source=lnms&tbm=isch&sa=X&ved=0ahUKEwio0Zm0oojTAhWCvrwKHTbaA8UQ_AUIBigB&biw=1434&bih=872#tbm=isch&q=google+rich+snippets&*) and rich cards are like [this](https://www.google.com.br/search?q=google+rich+cards&espv=2&source=lnms&tbm=isch&sa=X&ved=0ahUKEwio0Zm0oojTAhWCvrwKHTbaA8UQ_AUIBigB&biw=1434&bih=872#tbm=isch&q=google+rich+cards+example&*).

My references were: [Google Introduction to Structured Data](https://developers.google.com/search/docs/guides/intro-structured-data) and [Google Microdata](https://developers.google.com/schemas/formats/microdata)

Silence errors on Google tool is an attempt to get best SEO, moving to AMP will enable rich cards but it is a step beyond this PR. As told, I understand your concerns and agree; let´s keep current status.

BRs
